dwave/plugins/torch/samplers/block_spin_sampler.py

```diff
@@ -0,0 +1,348 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import annotations
+
+from abc import ABC, abstractmethod
+from collections import defaultdict
+from typing import TYPE_CHECKING, Callable, Hashable, Iterable, Literal, Optional
+
+import torch
+from torch import nn
+
+if TYPE_CHECKING:
+    from dwave.plugins.torch.models.boltzmann_machine import GraphRestrictedBoltzmannMachine as GRBM
+    from torch._prims_common import DeviceLikeType
+
+from dwave.plugins.torch.nn.functional import bit2spin_soft
+from dwave.plugins.torch.tensor import randspin
+
+__all__ = ["BlockSpinSampler"]
+
+
+class InvalidProposalAcceptanceCriteriaError(ValueError):
+    pass
+
+
+class NonPositiveIntegerError(ValueError):
+    pass
+
+
+class ShapeMismatchError(ValueError):
+    pass
+
+
+class NonSpinError(ValueError):
+    pass
+
+
+class InvalidColouringError(ValueError):
+    pass
+
+
+class BlockSpinSampler:
+    """A block-spin update sampler for graph-restricted Boltzmann machines.
+
+    Due to the sparse definition of GRBMs, some tedious indexing tricks are required to
+    efficiently sample in blocks of spins. Ideally, an adjacency list can be used, however,
+    adjacencies are ragged, making vectorization inapplicable.
+
+    Block-Gibbs and Block-Metropolis obey detailed balance and are ergodic methods at finite nonzero
+    temperature which, at fixed parameters, converge upon Boltzmann distributions. Block-Metropolis
+    allows higher acceptance rates for proposals (faster single-step mixing), but is non-ergodic in
+    the limit of zero or infinite temperature. Decorrelation from an initial condition can be slower.
+    Block-Gibbs represents best practice for independent sampling.
+
+    Args:
+        grbm (GRBM): The Graph-Restricted Boltzmann Machine to sample from.
+        crayon (Callable[Hashable, Hashable]): A colouring function; a function that maps a single
+            node of the `grbm` to its colour.
+        num_chains (int): Number of Markov chains to run in parallel.
+        initial_states (torch.Tensor | None): A tensor of +/-1 values of shape
+            (num_chains, grbm.n_nodes) Initial states of the Markov chains. If None, initial
+            states will be uniformly randomized with number of chains equal to ``num_chains``.
+            Defaults to None.
+        schedule (Iterable[Float]): The inverse temperature schedule.
+        proposal_acceptance_criteria (Literal["Gibbs", "Metropolis"]): The proposal acceptance
+            criterion used to accept or reject states in the Markov chain. Defaults to "Gibbs".
+        seed (Optional[int]): Random seed. Defaults to None.
+    Raises:
+        InvalidProposalAcceptanceCriteriaError: If the proposal acceptance criteria is not one of
+            "Gibbs" or "Metropolis".
+    """
+
+    def to(self, device: DeviceLikeType) -> None:
```

Comment:
I feel like this should either work in the same was torch's `to()` works (i.e., returns a copy of self on the new device) _or_ have a different name to do distinguish it (like `to_inplace()`).

Reply:
Added a `return self` statement. `nn.Module.to` move weights in-place.
```
from torch.nn import Linear
A = Linear(3, 2)
print(A.weight.device)
print(A.to("meta").weight.device)
print(A.weight.device)
```
Confusingly, `torch.Tensor` and `nn.Parameter` do not get moved in-place.